buildtools/AWS.Deploy.Tools.CI/AWS.Deploy.Tools.CI/CodeBuildFunctions.cs

```diff
@@ -0,0 +1,123 @@
+/// </summary>
+public class CodeBuildFunctions
+{
+    /// <summary>
+    /// The <see cref="GetAvailableTestRunner"/> function is responsible for checking available test runner accounts to run the AWS.Deploy.Tools CodeBuild PR check.
+    /// A list of IAM roles, representing the test runner accounts, is passed to the function. 
+    /// These roles are assumed and used to check if the CodeBuild CI project in the test runner account is currently running any jobs.
+    /// </summary>
+    /// <returns>The function will return the IAM role of the account that is not running any CodeBuild CI jobs.</returns>
+    /// <exception cref="ArgumentNullException">If the input passed to the function is invalid.</exception>
+    /// <exception cref="Exception">If no test runner account is available.</exception>
+    /// <exception cref="Exception">If a CodeBuild CI project is not found in the test runner account.</exception>
+    [LambdaFunction(Name = "GetAvailableTestRunner", Policies = "@AWSDeployToolsCIGetAvailableTestRunnerLambdaAssumeRolePolicy")]
+    public async Task<string> GetAvailableTestRunner([FromBody] GetAvailableTestRunnerInput input)
+    {
+        if (string.IsNullOrEmpty(input.Roles))
+        {
+            throw new ArgumentNullException(nameof(input.Roles));
+        }
+        if (string.IsNullOrEmpty(input.ProjectName))
+        {
+            throw new ArgumentNullException(nameof(input.ProjectName));
+        }
+        if (string.IsNullOrEmpty(input.Branch))
+        {
+            throw new ArgumentNullException(nameof(input.Branch));
+        }
+
+        var roles = input.Roles.Split(",").Select(x => x.Trim()).ToList();
+
+        if (!roles.Any())
+        {
+            throw new ArgumentNullException(nameof(input.Roles));
+        }
+
+        foreach (var role in roles)
+        {
+            var assumeRoleResponse =
+                await new AmazonSecurityTokenServiceClient().AssumeRoleAsync(
+                    new AssumeRoleRequest
+                    {
+                        RoleArn = role,
+                        RoleSessionName = "DeployToolTestRunner"
+                    }
+                );
+
+            var stsClient = new AmazonSecurityTokenServiceClient(assumeRoleResponse.Credentials);
+            var callerIdentity = await stsClient.GetCallerIdentityAsync(new GetCallerIdentityRequest());
+
+            var codeBuildClient = new AmazonCodeBuildClient(assumeRoleResponse.Credentials);
+
+            var batchGetProjectsResponse =
+                await codeBuildClient.BatchGetProjectsAsync(
+                    new BatchGetProjectsRequest
+                    {
+                        Names = new List<string> { input.ProjectName }
+                    }
+                );
+
+            if (!batchGetProjectsResponse.Projects.Any())
+            {
+                throw new Exception($"Could not find any project with the name '{input.ProjectName}' in account '{callerIdentity.Account}'.");
```

Comment:
If the project doesn't exist in a particular role should we just skip the role. I'm imagining when this function might be used for a collection of repositories and maybe we haven't created the appropriate codebuild job in all of our test accounts.

Reply:
I threw an exception purposefully to indicate that there is an issue with the infrastructure setup instead of failing silently and us having a hard time tracking why the job was not triggered. This Lambda function is reading IAM roles from the GitHub repo secrets. If the secrets have the IAM role, then I think it's safe to assume that the infrastructure was properly setup, since that IAM role is automatically generated by the CF template that I am using to create the Codebuild job as well.